src/expr.rs

```diff
@@ -35,7 +37,222 @@ pub enum Expr {
+                            "+" => {
+                                *self = Expr::Float(l + r);
+                            }
+                            "-" => {
+                                *self = Expr::Float(l - r);
+                            }
+                            "*" => {
+                                *self = Expr::Float(l * r);
+                            }
+                            "/" => {
+                                *self = Expr::Float(l / r);
+                            }
+                            "%" => {
+                                *self = Expr::Float(l % r);
+                            }
+                            "^" => {
+                                *self = Expr::Float(l.powf(r));
+                            }
+                            "=" => {
+                                *self = Expr::Bool(l == r);
+                            }
+                            "<=" => {
+                                *self = Expr::Bool(l <= r);
+                            }
+                            "<" => {
+                                *self = Expr::Bool(l < r);
+                            }
+                            ">=" => {
+                                *self = Expr::Bool(l >= r);
+                            }
+                            ">" => {
+                                *self = Expr::Bool(l > r);
+                            }
+                            "<>" => {
+                                *self = Expr::Bool(l != r);
+                            }
+                            _ => (),
+                        }
+                    }
+                }
+            }
+            _ => (),
+        }
+    }
+    /// Run CQL against a JSON Value
+    ///
+    ///  # Examples
+    ///
+    /// ```
+    /// use serde_json::{json, Value};
+    /// use cql2::Expr;
+    /// let item = json!({"properties":{"eo:cloud_cover":10, "datetime": "2020-01-01 00:00:00Z", "boolfield": true}});
+    /// let mut expr_json = json!(
+    ///     {
+    ///         "op" : ">",
+    ///         "args" : [
+    ///             {
+    ///                  "op": "+",
+    ///                  "args": [
+    ///                     {"property": "eo:cloud_cover"},
+    ///                     17
+    ///                  ]
+    ///             },
+    ///             2
+    ///         ]
+    ///     }
+    /// );
+    ///
+    ///
+    /// let mut expr: Expr = serde_json::from_value(expr_json).unwrap();
+    ///
+    ///
+    /// assert_eq!(true, expr.matches(&item).unwrap());
+    ///
+    ///
+    /// let mut expr2: Expr = "boolfield and 1 + 2 = 3".parse().unwrap();
+    /// assert_eq!(true, expr2.matches(&item).unwrap());
+    /// ```
+    pub fn matches(&self, j: &Value) -> Result<bool, ()> {
+        let mut e = self.clone();
+        e.reduce(j);
+        match e {
+            Expr::Bool(v) => Ok(v),
+            _ => Err(()),
+        }
```

Comment:
How often will we re-use `matches`? Can we make this consuming (taking `self`) and let the caller do a `clone` themselves? Cloning inside a simple method like this feels a bit off.